jre/java/java/util/stream/Collectors.java

```diff
@@ -170,6 +170,38 @@ public static Collector<CharSequence,?,String> joining(CharSequence delimiter) {
         downstream.finisher());
   }
 
+  public static <T, U, A, R> Collector<T, ?, R> flatMapping(Function<? super T,? extends Stream<?
+      extends U>> mapper, Collector<? super U, A, R> downstream) {
+    return new CollectorImpl<>(
+        downstream.supplier(),
+        (A a, T t) -> {
+            Stream<? extends U> stream = mapper.apply(t);
+            if (stream == null) {
+              return;
+            }
+            stream.forEach(u -> {
+              downstream.accumulator().accept(a, u);
+            });
+        },
+        downstream.combiner(),
+        downstream.finisher()
+    );
+  }
+
+  public static <T, A, R> Collector<T, ?, R> filtering(Predicate<? super T> predicate,
+                                                        Collector<? super T, A, R> downstream) {
+    return new CollectorImpl<>(
+        downstream.supplier(),
+        (a, t) -> {
+          if (predicate.test(t)) {
+            downstream.accumulator().accept(a, t);
+          }
+        },
+        downstream.combiner(),
+        downstream.finisher()
+    );
+  }
+
```

Comment:
Change is quite sizable already. Let's have a patch like "Add Stream.iterate APIs that takes predicate" and separate one for Collectors changes

Reply:
@treblereel There are still additional changes. Could you just keep the `iterate` API changes and updated PR to be "Add Stream.iterate APIs that takes predicate"?

Reply:
@gkdn ok, I did a separate PR https://github.com/google/j2cl/pull/230, I'll rebase this one, once https://github.com/google/j2cl/pull/230 will be merged.